server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java

```diff
@@ -73,33 +72,33 @@ public void handle(Request request, Response response) throws Exception {
     userSession.checkLoggedIn().checkPermission(GlobalPermissions.SYSTEM_ADMIN);
 
     String login = request.mandatoryParam(PARAM_LOGIN);
-    if (login.equals(userSession.getLogin())) {
-      throw new BadRequestException("Self-deactivation is not possible");
+    checkRequest(!login.equals(userSession.getLogin()), "Self-deactivation is not possible");
+
+    UserDto user;
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      user = dbClient.userDao().selectByLogin(dbSession, login);
+      checkFound(user, "User '%s' doesn't exist", login);
+
+      dbClient.userTokenDao().deleteByLogin(dbSession, login);
+      dbClient.userDao().deactivateUserByLogin(dbSession, login);
+      dbSession.commit();
     }
-    userUpdater.deactivateUserByLogin(login);
 
+    userIndexer.index();
     writeResponse(response, login);
   }
 
   private void writeResponse(Response response, String login) {
-    JsonWriter json = response.newJsonWriter().beginObject();
-    writeUser(json, login);
-    json.endObject().close();
-  }
-
-  private void writeUser(JsonWriter json, String login) {
-    json.name("user");
-    Set<String> groups = Sets.newHashSet();
-    DbSession dbSession = dbClient.openSession(false);
-    try {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      JsonWriter json = response.newJsonWriter().beginObject();
+      json.name("user");
+      Set<String> groups = Sets.newHashSet();
       UserDto user = dbClient.userDao().selectByLogin(dbSession, login);
-      if (user == null) {
-        throw new NotFoundException(format("User '%s' doesn't exist", login));
-      }
+      checkFound(user, "User '%s' doesn't exist", login);
```

Comment:
why starting to write JSON response and creating a Set if it's for potentially fail when user doesn't exist?

you could move line 96 and 97 to be beginning of the try